txnkv/txnsnapshot/snapshot.go

```diff
@@ -1025,6 +1078,17 @@ func (s *KVSnapshot) getResolveLockDetail() *util.ResolveLockDetail {
 	return s.mu.stats.resolveLockDetail
 }
 
+// SetPipelined sets the snapshot to pipelined mode.
+func (s *KVSnapshot) SetPipelined(ts uint64) {
+	s.isPipelined = true
+	// In pipelined mode, some locks are flushed into stores during the execution.
+	// If a read request encounters these pipelined locks, it'll be a situation where lock.ts == start_ts.
+	// In order to allow the snapshot to proceed normally, we need to skip these locks.
+	// Otherwise, the transaction will attempt to resolve its own lock, leading to a mutual wait with the primary key TTL.
+	// Currently, we skip these locks by resolvedLocks mechanism.
+	s.resolvedLocks.Put(ts)
```

Comment:
Should this be handled at the TiKV side when, it is known if the lock is a pipelined lock?
/cc @ekexium

Reply:
Here we just want to avoid txn resolve locks from itself, which becomes a deadlock.
The snapshot read in TiKV already handled the `resolved_locks` field, the behavior is same as ignoring non pipelined lock.

Reply:
Utilizing resolved_locks is enough for stage 1. Otherwise we'll have to modify the protocol which introduces unnecessary complexity.

Reply:
In tikv is it possible to decide if the lock should be read through or not by
```
if lock.type == pipelined_lock && lock.start_ts == read_ts {
   // Read the lock content if the read is union get
   // Skip the lock if the read is snapshot get
}
```
Not a big problem, maybe we could discuss later.